hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/handlers/marker/MarkerCheckerRunnable.java

```diff
@@ -0,0 +1,193 @@
+
+public class MarkerCheckerRunnable implements Runnable {
+  private static final Logger LOG = LogManager.getLogger(MarkerCheckerRunnable.class);
+
+  private MarkerHandler markerHandler;
+  private String markerDir;
+  private String basePath;
+  private FileSystem fs;
+  private AtomicBoolean hasConflict;
+  private long maxAllowableHeartbeatIntervalInMs;
+  private Set<HoodieInstant> oldInstants;
+
+  public MarkerCheckerRunnable(AtomicBoolean hasConflict, MarkerHandler markerHandler, String markerDir,
+                               String basePath, FileSystem fileSystem, long maxAllowableHeartbeatIntervalInMs,
+                               Set<HoodieInstant> oldInstants) {
+    this.markerHandler = markerHandler;
+    this.markerDir = markerDir;
+    this.basePath = basePath;
+    this.fs = fileSystem;
+    this.hasConflict = hasConflict;
+    this.maxAllowableHeartbeatIntervalInMs = maxAllowableHeartbeatIntervalInMs;
+    this.oldInstants = oldInstants;
+  }
+
+  @Override
+  public void run() {
+    try {
+      if (!fs.exists(new Path(markerDir))) {
+        return;
+      }
+
+      HoodieTimer timer = new HoodieTimer().startTimer();
+      Set<String> currentInstantAllMarkers = markerHandler.getAllMarkers(markerDir);
+      Path tempPath = new Path(basePath + Path.SEPARATOR + HoodieTableMetaClient.TEMPFOLDER_NAME);
+
+      List<Path> instants = MarkerUtils.getAllMarkerDir(tempPath, fs);
+      List<String> candidate = getCandidateInstants(instants, markerDirToInstantTime(markerDir));
+      Set<String> tableMarkers = candidate.stream().flatMap(instant -> {
+        return MarkerUtils.readTimelineServerBasedMarkersFromFileSystemLocally(instant, fs).stream();
+      }).collect(Collectors.toSet());
+
+      Set<String> currentFileIDs = currentInstantAllMarkers.stream().map(this::makerToPartitionAndFileID).collect(Collectors.toSet());
+      Set<String> tableFilesIDs = tableMarkers.stream().map(this::makerToPartitionAndFileID).collect(Collectors.toSet());
+
+      currentFileIDs.retainAll(tableFilesIDs);
+
+      if (!currentFileIDs.isEmpty() || hasCommitConflict(currentInstantAllMarkers, basePath)) {
+        LOG.warn("Conflict writing detected based on markers!\n"
+            + "Conflict markers: " + currentInstantAllMarkers + "\n"
+            + "Table markers: " + tableMarkers);
+        hasConflict.compareAndSet(false, true);
+      }
+      LOG.info("Finish batch marker checker in " + timer.endTimer() + " ms");
+
+    } catch (IOException e) {
+      throw new HoodieIOException("IOException occurs during checking marker conflict");
+    }
+  }
+
+  /**
+   * Get Candidate Instant to do conflict checking:
+   * 1. Skip current writer related instant(currentInstantTime)
+   * 2. Skip all instants after currentInstantTime
+   * 3. Skip dead writers related instants based on heart-beat
+   * @param instants
+   * @return
+   */
+  private List<String> getCandidateInstants(List<Path> instants, String currentInstantTime) {
```

Comment:
Can we adapt the common logic from `ConflictResolutionStrategy` instead of reinventing similar logic?

Reply:
Also Thanks for your reviewing here! Really appreciate!